pretty: miscellaneous formatting improvements

knz opened the pull request:
This is not yet in the domain of "bikeshedding" -- more alike to "general clean-up"

Cumulative perf improvements:

```
name           old time/op    new time/op    delta
PrettyData-16     4.48s ±21%     1.77s ±26%  -60.44%  (p=0.008 n=5+5)
PrettyData-16    3.20GB ± 2%    0.62GB ± 1%  -80.56%  (p=0.008 n=5+5)
PrettyData-16     10.5M ± 0%      4.2M ± 0%  -59.67%  (p=0.008 n=5+5)
```

knz committed:
pretty: improve the formatting of CASE...END

The formatting on long lines remains unchanged.

When width-constrained, before:

```
---------------- width -----------------|
SELECT                                --|
    CASE x			      --|
    WHEN 'a'			      --|
    THEN 'aaaa'			      --|
    WHEN 'b'			      --|
    THEN 'bbbb'			      --|
    ELSE 'ccccc'		      --|
    END                               --|
```

After:

```
---------------- width -----------------|
SELECT                                --|
    CASE x			      --|
    WHEN 'a' THEN 'aaaa'	      --|
    WHEN 'b' THEN 'bbbb'	      --|
    ELSE 'ccccc'		      --|
    END                               --|
```

i.e. this patch advantages placing the WHEN and THEN on the same line.

Release note: None

knz committed:
pretty: improve the formatting of joins

On long lines, before:

```
--------------------------------- width ---------------------------------------|
SELECT *                                                                     --|
FROM                                                                         --|
    kv INNER JOIN blah ON kv.x = blah.y INNER JOIN foo ON blah.x = foo.z     --|
        INNER JOIN bzz ON bzz.z = foo.z,                                     --|
    blah INNER JOIN blih ON blah.x = blih.y                                  --|
```

After:

```
--------------------------------- width ---------------------------------------|
SELECT *                                                                     --|
FROM                                                                         --|
    kv                                                                       --|
    INNER JOIN blah ON kv.x = blah.y                                         --|
    INNER JOIN foo ON blah.x = foo.z                                         --|
    INNER JOIN bzz ON bzz.z = foo.z,                                         --|
    blah                                                                     --|
    INNER JOIN blih ON blah.x = blih.y                                       --|
```

i.e. each join operand on its own line.

Width constrained, before:

```
----------- width ------------|
SELECT *                    --|
FROM                        --|
    kv                      --|
        INNER JOIN blah     --|
        ON kv.x = blah.y    --|
        INNER JOIN foo      --|
        ON blah.x = foo.z   --|
        INNER JOIN bzz      --|
        ON bzz.z = foo.z    --|
```

After:

```
----------- width ------------|
SELECT *                    --|
FROM                        --|
    kv                      --|
    INNER JOIN blah         --|
    ON kv.x = blah.y        --|
    INNER JOIN foo          --|
    ON blah.x = foo.z       --|
    INNER JOIN bzz          --|
    ON bzz.z = foo.z        --|
```

i.e. the various join operands are vertically aligned; there is no
special alignment emphasis on the first operand of a join.

Release note: None

knz committed:
pretty: improve the formatting of conjunctions and disjunctions

Two fixes:

1) multi-operand expressions are flattened.

Before:

```
SELECT (1 AND 2) AND 3
```

After:

```
SELECT 1 AND 2 AND 3
```

2) the operands are aligned.

Before:

```
SELECT
    (
        (
            (
                (
                    (
                        (
                            (1 AND 2)
                            AND 3
                        )
                        AND 4
                    )
                    AND 5
                )
                AND 6
            )
            AND 7
        )
        AND 8
    )
    AND 9
```

After:

```
SELECT
    1
    AND 2
    AND 3
    AND 4
    AND 5
    AND 6
    AND 7
    AND 8
    AND 9
```

Note that this is careful to preserve operator precedence; compare:

```
-- this input:
select 1 and 2 and 3 or 4 and 5 and 6 or 7 and 8 and 9
-- is transformed to this:
SELECT
    (1 AND 2 AND 3)
    OR (4 AND 5 AND 6)
    OR (7 AND 8 AND 9)
```

```
-- this input:
select 1 or 2 or 3 and 4 or 5 or 6 and 7 or 8 or 9
-- is transformed to this:
SELECT
    1
    OR 2
    OR (3 AND 4)
    OR 5
    OR (6 AND 7)
    OR 8
    OR 9
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27225)
<!-- Reviewable:end -->

knz committed:
pretty: improve the formatting of set operators

Prior to this patch the following statement was surprisingly (=
incorrectly) formatted:

```
-------- width ----------|
SELECT *               --|
        FROM kv	       --|
        WHERE v = 3    --|
    UNION	       --|
        SELECT *       --|
        FROM kv	       --|
        WHERE k = 2    --|
UNION		       --|
    SELECT *	       --|
    FROM kv	       --|
    WHERE (k + v) < 2  --|
```

Note how the code was meant to indent both the left and right operand
to UNION relative to the operator, but yet the first SELECT line is
yanked to the left.

The root of the problem is that the formatting algorithm does not
support indenting a nested document if there is nothing before
it. This is the key difference between P. Wadler's idea (used here)
and the original algorithm by J. Hugues, see section 5 of Wadler's
note at
https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf.

This patch fixes the issue by de-indenting the left operand. This adds
the benefit that an expression of the form (a union b) union c has its
b and c operands properly aligned.

After:

```
-------- width ----------|
SELECT *               --|
FROM kv		       --|
WHERE v = 3	       --|
UNION		       --|
    SELECT *	       --|
    FROM kv	       --|
    WHERE k = 2	       --|
UNION		       --|
    SELECT *	       --|
    FROM kv	       --|
    WHERE (k + v) < 2  --|
```

Release note: None

knz committed:
pretty: improve the formatting of binary scalar operators

Before this patch, under width constraints:

```
SELECT
    10
    +
    (
        count
        *
        60
        *
        24
    )
```

After:

```
SELECT
    10
    + (
        count
        * 60
        * 24
    )
```

i.e. yank the right operand on the same line as the operator, for
readability.

Release note: None

knz committed:
pretty: introduce an option to remove extraneous parentheses

When the "simplify" option is set, parentheses around scalar
sub-expressions are removed automatically when the result is
grammatically unambiguous and equivalent to the original syntax.

Before:

```
SELECT (a + b) * (c * d);
```

After:

```
SELECT (a + b) * c * d;
```

Or, more convincingly, from before:

```
SELECT
    (
        (
            (
                (
                    (
                        (
                            ((CASE WHEN rolcanlogin THEN '-- User: ' ELSE '-- Role: ' END || quote_ident(rolname)) || e'\\n-- DROP ')
                            || CASE WHEN rolcanlogin THEN 'USER ' ELSE 'ROLE ' END
                        )
                        || quote_ident(rolname)
                    )
                    || e';\\n\\nCREATE '
                )
                || CASE WHEN rolcanlogin THEN 'USER ' ELSE 'ROLE ' END
            )
            || quote_ident(rolname)
        )
        || e' WITH\\n  '
    )
```

To after:

```
SELECT
    CASE WHEN rolcanlogin THEN '-- User: ' ELSE '-- Role: ' END || quote_ident(rolname) || e'\\n-- DROP ' || CASE WHEN rolcanlogin THEN 'USER ' ELSE 'ROLE ' END
    || quote_ident(rolname)
    || e';\\n\\nCREATE '
    || CASE WHEN rolcanlogin THEN 'USER ' ELSE 'ROLE ' END
    || quote_ident(rolname)
    || e' WITH\\n  '
```

Release note: None

madelynnblue reviewed (approved):
These are great. Thanks.

knz committed:
pretty: remove ROWS FROM when unambiguous

If there is just one operand in ROWS FROM, and it is a function
application, then it can be elided.

Release note: None

knz committed:
pretty: improve the formatting of FROM clauses

When simplifying, this strips superfluous parentheses in FROM clauses.

Before:

```
SELECT *
FROM
    (a CROSS JOIN b),
    (
        a
            CROSS JOIN
                (
                    c
                        CROSS JOIN
                            d
                )
    ),
    d,
    (TABLE e ORDER BY f)
```

After:

```
SELECT *
FROM
    a
    CROSS JOIN b,
    a
    CROSS JOIN
        (c CROSS JOIN d),
    d,
    (TABLE e ORDER BY f)
```

Release note: None

knz committed:
pretty: improve the formatting of comparison expressions

This indents the right operand if it needs to be wrapped, and
conditionally removes some extraneous parentheses.

Before:

```
------ width -------|
SELECT            --|
    1 = 1,	  --|
    (1 + 2)	  --|
    = (3 * foo(x)),
    ((x))	  --|
    = (y::INT),	  --|
    (y[123])	  --|
    = (min(z))    --|
```

After:

```
------ width -------|
SELECT            --|
    1 = 1,	  --|
    (1 + 2)       --|
    = (3 * foo(x)),
    x = y::INT,   --|
    y[123] = min(z)
```

Release note: None

knz committed:
pretty: improve the formatting of UPDATE

Before:

```
UPDATE foo
SET
    x
    =
    (3 + 2),
    (y, z)
    =
    (
        4
        +
        min(1 - 2),
        123
    )
```

After:

```
UPDATE foo
SET
    x = 3 + 2,
    (y, z)
        = (
            4
            + min(
                    1
                    - 2
                ),
            123
        )
```

i.e. the assignments are shifted to the left and, if possible, the LHS
and RHS of each assignment are grouped on the same line.

Release note: None

knz commented:
and there we go

bors r+

mjibson commented on pkg/util/pretty/document.go:
```diff
@@ -106,6 +106,24 @@ func Concat(a, b Doc) Doc {
+// Fill concatenates a list of docs with spaces if possible,
+// otherwise with newlines.
+// See linked paper pp 14-15.
+// func Fill(d ...Doc) Doc {
```
Was this debug work? Probably wasn't supposed to be in this commit but whatever.

knz replied:
I had tried it out but then commented it out to please the linter. I think it's a nice function but we haven't used it. I was happy while writing this to see how easy it was to reproduce the algorithms!

mjibson commented on pkg/sql/sem/tree/pretty.go:
```diff
@@ -799,7 +799,11 @@ func (node *UpdateExpr) doc(p PrettyCfg) pretty.Doc {
+	if p.Simplify {
+		e = StripParens(e)
+	}
+	return pretty.Group(p.nestName(d, pretty.ConcatSpace(pretty.Text("="), p.Doc(e))))
```
nestName calls Group on its outer Doc, so the call to Group here is extraneous.

knz replied:
I see you fixed it. Thanks!

madelynnblue reviewed (commented):
I approved this PR and then it changed a bunch, so here are the rest of my comments.

knz commented on pkg/sql/sem/tree/pretty.go:
```diff
@@ -252,24 +269,57 @@ func (node *Exprs) doc(p PrettyCfg) pretty.Doc {
+	case *BinaryExpr:
+		parenPrio := binaryOpPrio[op]
+		childPrio := binaryOpPrio[te.Operator]
+		if te.Operator == op && childPrio <= parenPrio {
```
Thanks I had broken this inadvertently. See #27255.
